src/null-safety/unsound-null-safety.md

```diff
@@ -0,0 +1,186 @@
+---
+title: Unsound null safety
+description: Mixing language versions lets you migrate to null safety at your own pace, with some of the benefits of null safety.
+---
+
+An app can use some libraries that
+are [null safe][] and some that aren't.
+These apps — called **mixed-version programs** —
+execute with **unsound null safety**.
+
+[null safe]: /null-safety
+
+Mixing language versions lets you migrate to null safety at your own pace.
+It also frees package maintainers to migrate their code,
```

Comment:
I think this is more or less from @munificent (although I probably changed it, perhaps changing the meaning), so maybe he can help us clear it up.

Reply:
@munificent can you clarify here? Here are the parts that confuse me:

> Mixing [language versions][] lets you migrate to null safety at your own pace.

Yup, this part I agree with. Perhaps "own pace" is a bit imprecise; I think the core value in mixed versions is that you can migrate parts of a package or app rather than all of it at once?

> It also frees package maintainers to migrate their code, with the knowledge that even legacy users can get new bug fixes and other improvements.

This part I don't get. How does mixed mode "free" the maintainer? Is this about it enabling them to migrate "out of order", i.e. migrate before their dependencies have migrated to null safety?

> However, mixed-version programs don't get all the advantages that null safety can bring.

Yup, this is good.

Reply:
Ah, I see what you mean. Mixed-version comes into play in two use cases:

1. A developer migrates a package, all the way to sound nulls safety. Because of mixed-version support, this package can be used _not only_ by apps that themselves use null safety, but also by "legacy" apps that don't yet use null safety. This is a form of mixed version, but it's perhaps easier to understand as backwards compatibility.

   This is a *good way* of using null safety.

2. A developer has a package foo with two deps, bar and baz. Bar has migrated to null safety, but baz hasn't. The developer doesn't feel like waiting on baz, so they migrate their package foo out of order. This is also a form of mixed version.

   This is a potentially more troublesome way of using null safety. You can do it, but we should recommend against it. The core issue is that once that last dependency baz migrates, the run-time checking from soundness might be breaking to consumers of **bar**.

Did I get that right?

Reply:
I modified the text a bit to simplify it and focus on the case of migrating from the leaf dependencies. ptal

PS: I should probably add the info from https://github.com/dart-lang/site-www/pull/2739#discussion_r526430314, but maybe not in the first published version.